plugins/Goals/DataTable/Filter/CalculateConversionPageRate.php

```diff
@@ -0,0 +1,119 @@
+<?php
+/**
+ * Matomo - free/libre analytics platform
+ *
+ * @link https://matomo.org
+ * @license http://www.gnu.org/licenses/gpl-3.0.html GPL v3 or later
+ *
+ */
+namespace Piwik\Plugins\Goals\DataTable\Filter;
+
+use Piwik\API\Request;
+use Piwik\DataTable\BaseFilter;
+use Piwik\DataTable;
+use Piwik\Metrics;
+use Piwik\Piwik;
+use Piwik\Site;
+
+class CalculateConversionPageRate extends BaseFilter
+{
+    /**
+     * Constructor.
+     *
+     * @param DataTable $table The table to eventually filter.
+     */
+    public function __construct($table)
+    {
+        parent::__construct($table);
+    }
+
+    /**
+     * @param DataTable $table
+     */
+    public function filter($table)
+    {
+
+        // Find all goal ids used in the table and store in an array
+        $goals = [];
+        foreach ($table->getRowsWithoutSummaryRow() as $row) {
+            if (isset($row[Metrics::INDEX_GOALS])) {
+                foreach ($row[Metrics::INDEX_GOALS] as $goalIdString => $metrics) {
+                    $goals[$goalIdString] = $goalIdString;
+                }
+            }
+        }
+
+        // Get the total top-level conversions for the goals in the table
+        $goalTotals = $this->getGoalTotalConversions($table, $goals);
+        if (count($goalTotals) === 0) {
+            return;
+        }
+
+        // Walk the rows and populate the nb_conversions_page_rate with nb_conversions_page_uniq / $goalTotals[goal id]
+        foreach ($table->getRowsWithoutSummaryRow() as &$row) {
+            if (isset($row[Metrics::INDEX_GOALS])) {
+                foreach ($row[Metrics::INDEX_GOALS] as $goalIdString => $metrics) {
+                    if (isset($row[Metrics::INDEX_GOALS][$goalIdString][Metrics::INDEX_GOAL_NB_CONVERSIONS_PAGE_UNIQ])) {
+
+                        $rate = Piwik::getQuotientSafe(
+                                $row[Metrics::INDEX_GOALS][$goalIdString][Metrics::INDEX_GOAL_NB_CONVERSIONS_PAGE_UNIQ],
+                                $goalTotals[$goalIdString],
+                                3
+                            );
+                        // Prevent page rates over 100% which can happen when there are subpages
+                        if ($rate > 1) {
+                            $rate = 1;
+                        }
+
+                        $row[Metrics::INDEX_GOALS][$goalIdString][Metrics::INDEX_GOAL_NB_CONVERSIONS_PAGE_RATE] = $rate;
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * Get the conversions total for each goal in the top level datatable
+     *
+     * @param DataTable $table
+     * @param array $goalIds
+     * @return array
+     */
+    private function getGoalTotalConversions(DataTable $table, array $goalIds)
```

Comment:
We should also use a return type hint for methods where it's possible.
```suggestion
    private function getGoalTotalConversions(DataTable $table, array $goalIds): array
```